### DISCUSSION TITLE
Get FoCo Database Rearchitecture

### DISCUSSION THREAD

**Discussion by penoptech** [Get-Your/Get-Your#95] [General]
@TiCampbell, you've probably already thought about this, but here was something I had a thought on - related to the rearchitecture.

I was going through and updating the app layer for the database rearchitecture. I started thinking about how once the updates are completed how we'll move everything into the new Django `platform` application. Since we're going to upgrade Django to 4.1 (or higher) and we've talked about linting all of the code, I think it would make sense to create an orphan branch that would have no reference to any existing branch in the repo. Doing this would let us accomplish a few things:

- Start Clean with Django 4
- Lint files
- Pivot to using the new `platform` Django application
- Consolidate static files into the `platform` application

I think doing this would making the environment we're working in cleaner. We won't have to work around existing and new code and database schemas. It can just be new and only new, making reasoning about everything much easier.

One thing to consider though (and I think it might already be done) is we'll have to have ETL scripts to load the data into the new database.

I think the we should finish the work we've done in the `db-rearch-p1` branch (and maybe `db-rearch-p2`) first. That way we're essentially just copying and pasting code over to the new `orphan` branch.

There's another thought here too related to making the project as "white label" as possible. Should we rename the code repository and database? If other municipalities we're to adopt it, I wonder if it would make sense to adopt you `platform` naming across everything (repository, database, etc.).

My ultimately goal here is to start off the V2 code base on the cleanest foot possible.

**Comment by TiCampbell** [comment]
@penoptech This all sounds great and I'm glad we're working in a brand-new repo now, with blank `dev` and `main` (although I also brought over those branches from the `GetFoco` repo as `dev_getfoco` and `main_getfoco`, respectively, for reference).

I added #97 to the tracker for ETL design (it's completed for phase0 > phase1 changes, but phase2 is turning out to be a decent-sized change as well). I like your idea of having it all the same label (after all, our server is named for Get FoCo, which should be enough), so I created `platform_dev` on the DEV server and `platform_stage` and `platform_prod` on the PROD server (and added #98 to set up the permissions).

**Comment by penoptech** [comment]
@TiCampbell - Was working through updating the quick apply functionality to support the database rearchitecture and noticed we don't have anything in the new schema to store something the like the `spin_privacy_acknowledgement`. I'm sure it can be normalized more than how it currently is in `GetFoco` repo.

**Comment by TiCampbell** [comment]
@penoptech Great call. I'll add it to my list. I probably will be updating the schema this afternoon with the eligibility-program-based income selection anyway...I'll update here when I make the change(s).

**Comment by penoptech** [comment]
@TiCampbell - I've been holding off on updating the address part of the application, because it scares me. I finally faced my fears and started to work on it this weekend. Below are my thoughts on the implementation. I think the big one that I wanted to get your opinion on was adding an additional column to the `platform_addressrd` table. ~~It seems like we'd want to check this table at some point in the process to see if an address already exists, but the only thing that came to me was creating a hash of the address and storing in it with the address, so we could query the address from the `platform_addressrd` table. What do you think?~~ (doh!) I see the composite key you made on the address

[Blank diagram (1).pdf](https://github.com/Get-Your/Get-Your/files/11305199/Blank.diagram.1.pdf)

**Comment by TiCampbell** [comment]
@penoptech Ha that's fair, this is the biggest singular change to the functionality.

I did put that composite key on `platform_addressrd` to account for this, but I like your hash idea better... the lookup should be quicker than a composite text key. Let's go with a SHA-1 hash and name it `platform_addressrd.address_sha1` (I'll update #78 with the changes). I mocked up some code below from my standard hash functions; I only included the data fields in the hash, so we'll probably want to recalculate the metadata fields (`is_in_gma`, `is_city_covered`, `has_connexion`) each time the address is accessed.

I also agree with you on leaving in the 'in progress' addresses. I included the `is_verified` Boolean from the v1 app as the final flag for whether the address has completed verification, so I think we can keep the same (or similar) workflow.

    def hash_address(
            address_dict: dict,
            ) -> str:
        """
        Create a SHA-1 hash from existing address values.
    
        Parameters
        ----------
        address_dict : dict
            Dictionary of user-entered address fields.
    
        Returns
        -------
        str
            String representation of SHA-1 address hash. SHA-1 hash is 160 bits;
            written as hex for 40 characters.
    
        """
        
        import hashlib
        
        # Explicitly define address field order
        keyList = [
            'address1',
            'address2',
            'city',
            'state',
            'zip_code',
            ]
    
        # Concatenate string representations of each value in sequence.
        # If key DNE, use blank string.
        concatVals = ''.join([str(address_dict[key]) if key in address_dict.keys() else '' for key in keyList])
        
        # Return SHA-1 hash of the concatenated strings
        return(
            hashlib.sha1(
                bytearray(concatDictVals, 'utf8')
                ).hexdigest()
            )

**Comment by TiCampbell** [comment]
@penoptech I'd like to complete the TODO in your diagram to support users outside the GMA as *part* of the v2 release rather than after. That was my idea behind adding the `req_...` fields in the `iqprogramrd` table now, in hopes of not wasting effort on refactoring the current GMA check and going straight to a program-level filter. What do you think?

**Comment by TiCampbell** [comment]
@penoptech *Confirmed*: we're good to go to get rid of the GAHI dropdown!!

As discussed, we'll replace it with the minimum AMI threshold of the selected programs (e.g. `min( Program0.platform_eligibilityprogramrd.ami_threshold, Program1.platform_eligibilityprogramrd.ami_threshold, ...)`).

I'll update the #78 ERD at some point, but for now below are our notes from our meeting. The changes are:
- Remove `platform_ami` table
- In `platform_household`
  - Remove `ami_year`
  - Remove `ami_range_min`? This might be helpful for analytics, but I can't think of any specific use case
  - Add field (and question) for whether the applicant rents or owns (what should this field be named?)
  - Is there a clearer name for `ami_range_max` now?

![image](https://user-images.githubusercontent.com/44819630/235168308-c1d7bce7-ddae-4adf-8512-99ed603e7299.png)

**Comment by TiCampbell** [comment]
We also discussed keeping the 'TODO' to support users outside the GMA as well as switching to an address hash as tasks *after* the initial v2 release; I added them with notes that they're future changes (#82 and #121, respectively).

**Comment by TiCampbell** [comment]
Closing as (long-since) complete. No notes to move elsewhere.
